[Intel GPU] allow_tf32 for oneDNN backend - XPU part

ZhiweiYan-96 opened the pull request:
# Motivation
Add context variable `torch.bachend.mkldnn.allow_tf32` to control tf32 computation in convolution kernels at XPU side.  The tf32 data type is beneficial to improve the performance of deep learning workloads during training/inference. Current PR uses the [oneDNN API fpmath_mode](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_fpmath_mode.html#the-floating-point-math-mode-attribute) to trigger the tf32 acceleration in convolution kernels.

# Valiadation
* ut to test context variable 
`python test/xpu/test_conv.py -k test_mkldnn_allow_tf32_get_set`

* Runtime exemplification
```
onednn_verbose,primitive,exec,gpu:0,convolution,jit:ir,forward_training,src_f32::blocked:abcd::f0 wei_f32::blocked:abcd::f0 bia_f32::blocked:a::f0 dst_f32::blocked:abcd::f0,attr-scratchpad:user attr-fpmath:tf32,alg:convolution_direct,mb20_ic16oc33_ih50oh24kh3sh2dh0ph0_iw100ow49kw3sw2dw0pw0,0.649902
onednn_verbose,primitive,exec,gpu:0,convolution,jit:ir,forward_training,src_f32::blocked:abcd::f0 wei_f32::blocked:abcd::f0 bia_f32::blocked:a::f0 dst_f32::blocked:abcd::f0,attr-scratchpad:user attr-fpmath:tf32,alg:convolution_direct,mb20_ic33oc33_ih24oh24kh3sh1dh0ph1_iw49ow49kw3sw1dw0pw1,0.151855
onednn_verbose,primitive,exec,gpu:0,convolution,jit:ir,backward_data,src_f32::blocked:abcd::f0 wei_f32::blocked:abcd::f0 bia_undef::undef::: dst_f32::blocked:abcd::f0,attr-scratchpad:user attr-fpmath:tf32,alg:convolution_direct,mb20_ic33oc33_ih24oh24kh3sh1dh0ph1_iw49ow49kw3sw1dw0pw1,0.167969
onednn_verbose,primitive,exec,gpu:0,convolution,jit:ir,backward_weights,src_f32::blocked:abcd::f0 wei_f32::blocked:abcd::f0 bia_f32::blocked:a::f0 dst_f32::blocked:abcd::f0,attr-scratchpad:user attr-fpmath:tf32,alg:convolution_direct,mb20_ic33oc33_ih24oh24kh3sh1dh0ph1_iw49ow49kw3sw1dw0pw1,0.26709
onednn_verbose,primitive,exec,gpu:0,convolution,jit:ir,backward_weights,src_f32::blocked:abcd::f0 wei_f32::blocked:abcd::f0 bia_f32::blocked:a::f0 dst_f32::blocked:abcd::f0,attr-scratchpad:user attr-fpmath:tf32,alg:convolution_direct,mb20_ic16oc33_ih50oh24kh3sh2dh0ph0_iw100ow49kw3sw2dw0pw0,0.219971

```
According to the field `fpmath:tf32` in verbose, we could see that, current context setting utils could successfully trigger tf32 computation in conv forward/backward_data/backward_weights kernels.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137570



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @EikanWang @fengyuan14 @guangyey

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

EikanWang commented on aten/src/ATen/Context.cpp:
```diff
@@ -121,6 +121,14 @@ void Context::setAllowTF32CuDNN(bool b) {
   allow_tf32_cudnn = b;
 }
 
+bool Context::allowTF32Mkldnn() const {
```
@yanbing-j , should we use "onednn" rather than "mkldnn"? Because, I noticed you maintain a huge ghstack to replace mkldnn with onednn.

yanbing-j replied:
@EikanWang Is this a new API which conains 'mkldnn'? If so, you should use 'onednn' directly. Otherwise, I will change it to 'onednn' in my ghstack when this PR is merged.
For this new added API, you can use 'onednn' directly. For other existing API usage, you can use 'mkldnn', and I will rebase for that.

EikanWang replied:
Yes, this is a new API. @ZhiweiYan-96 let's refine the API name by replacing `mkldnn` with `onednn`. Meanwhile, pls. link @yanbing-j's PR to help reviewers get the background.

yanbing-j replied:
ghstack starts from https://github.com/pytorch/pytorch/pull/133289.

ZhiweiYan-96 replied:
sure, I will change the word, and thanks for your kindly clarification.

ZhiweiYan-96 replied:
I have changed word "mkldnn" need to be changed in the PR, thanks for your comments.

EikanWang commented on test/xpu/test_conv.py:
```diff
@@ -1264,6 +1264,17 @@ def test_channels_last_ouput_stride(self, device, dtype):
             assert_size_stride(out, (2, 512, 7, 7), (25088, 1, 3584, 512))
 
+    @onlyXPU
+    def test_mkldnn_allow_tf32_get_set(self, device):
```
Is `device` a useless variable here?

ZhiweiYan-96 replied:
Yes, it is now removed in newest commit, thanks.

EikanWang commented on torch/csrc/Module.cpp:
```diff
@@ -888,6 +888,25 @@ PyObject* THPModule_setDeterministicAlgorithms(
+  HANDLE_TH_ERRORS
+  TORCH_CHECK(
+      PyBool_Check(arg),
+      "set_allow_tf32_cublas expects a bool, "
```
`set_allow_tf32_cublas` -> `set_allow_tf32_mkldnn`.

ZhiweiYan-96 replied:
Thanks for the reminding, have modified

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

guangyey commented on aten/src/ATen/Context.cpp:
```diff
@@ -121,6 +121,14 @@ void Context::setAllowTF32CuDNN(bool b) {
   allow_tf32_cudnn = b;
 }
 
+bool Context::allowTF32Onednn() const {
```
```suggestion
bool Context::allowTF32OneDNN() const {
```
Is this better?

ZhiweiYan-96 replied:
I perfer we align the style with change from CPU here https://github.com/pytorch/pytorch/pull/133315? @yanbing-j  @guangyey

guangyey commented on aten/src/ATen/Context.cpp:
```diff
@@ -121,6 +121,14 @@ void Context::setAllowTF32CuDNN(bool b) {
+  return allow_tf32_onednn;
+}
+
+void Context::setAllowTF32Onednn(bool b){
```
```suggestion
void Context::setAllowTF32OneDNN(bool b){
```

guangyey commented:
FYI: be aware of https://github.com/pytorch/pytorch/pull/125888

ZhiweiYan-96 commented:
> FYI: be aware of #125888

Thanks for you kindly reminding, I need check this PR before any further actions.

ZhiweiYan-96 commented:
> > FYI: be aware of #125888
> 
> Thanks for you kindly reminding, I need check this PR before any further actions.

@guangyey  @EikanWang , I have read the https://github.com/pytorch/pytorch/pull/125888. Seems the PR is not finalized, we may firstly align with the currently API (`torch.backedn.mklddn.allow_tf32`) at present and move to the new design once it is finalized?

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

guangyey commented on torch/csrc/Module.cpp:
```diff
@@ -920,6 +920,29 @@ static PyObject* THPModule_setDeterministicAlgorithms(
+  HANDLE_TH_ERRORS
+  TORCH_CHECK(
+      PyBool_Check(arg),
+      "set_allow_tf32_onednn expects a bool, "
```
```suggestion
      "_set_onednn_allow_tf32 expects a bool, "
```

ZhiweiYan-96 replied:
Thanks for you reminding, the code has been changed here

guangyey commented:
Please fix the typo.

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

EikanWang commented on aten/src/ATen/native/mkldnn/xpu/detail/Conv.cpp:
```diff
@@ -120,6 +120,12 @@ sycl::event convolution(
   }
 #endif
 
+  auto& ctx = at::globalContext();
+  bool allow_tf32 = ctx.allowTF32Onednn();
+  if (allow_tf32) {
+    pattr.set_fpmath_mode(dnnl::fpmath_mode::tf32);
+  }
+
```
Pls. extract the logic as a utility function.

```C++
inline void apply_tf32_if_allowed(dnnl::primitive_attr& primitive_attr) {
  auto& ctx = at::globalContext();
  bool allow_tf32 = ctx.allowTF32Onednn();
  if (allow_tf32) {
    pattr.set_fpmath_mode(dnnl::fpmath_mode::tf32);
  }
}
```

ZhiweiYan-96 replied:
Thanks for suggestion, the utils have been extracted at https://github.com/pytorch/pytorch/pull/137570/files#diff-e2daddb2823a7fbde7e1c8944ac3d266c1270580b73e8d18ae333149540aa875R453

EikanWang commented on torch/_C/__init__.pyi.in:
```diff
@@ -1171,6 +1171,8 @@ def _get_cudnn_deterministic() -> _bool: ...  # THPModule_deterministicCuDNN
 def _set_cudnn_deterministic(arg: _bool) -> None: ...  # THPModule_setDeterministicCuDNN
 def _get_mkldnn_deterministic() -> _bool: ...  # THPModule_deterministicMkldnn
 def _set_mkldnn_deterministic(arg: _bool) -> None: ...  # THPModule_setDeterministicMkldnn
+def _get_onednn_allow_tf32() -> _bool: ... # THPModule_allowTF32Onednn
```
Pls. align with prior naming style. `Onednn` -> `OneDNN`

EikanWang commented on torch/csrc/Module.cpp:
```diff
@@ -920,6 +920,29 @@ static PyObject* THPModule_setDeterministicAlgorithms(
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject* THPModule_setAllowTF32Onednn(
```
`THPModule_setAllowTF32Onednn` -> `THPModule_setAllowTF32OneDNN`

EikanWang commented on torch/csrc/Module.cpp:
```diff
@@ -920,6 +920,29 @@ static PyObject* THPModule_setDeterministicAlgorithms(
+  END_HANDLE_TH_ERRORS
+}
+
+static PyObject* THPModule_allowTF32Onednn(
```
ditto

EikanWang commented on aten/src/ATen/Context.cpp:
```diff
@@ -122,6 +122,14 @@ void Context::setAllowTF32CuDNN(bool b) {
+  return allow_tf32_onednn;
+}
+
+void Context::setAllowTF32Onednn(bool b){
```
@ZhiweiYan-96 , Guanye's comment is marked as resolved. Do you mean it is not applicable?

ZhiweiYan-96 replied:
No, the comments here is same as the one above. I responded in the above comments. The naming here is aligned to @yanbing-j 's style.

EikanWang commented on aten/src/ATen/Context.cpp:
```diff
@@ -122,6 +122,14 @@ void Context::setAllowTF32CuDNN(bool b) {
+  return allow_tf32_onednn;
+}
+
+void Context::setAllowTF32Onednn(bool b){
```
Pls. rename `b` as readable.

ZhiweiYan-96 replied:
We use `b`  here to align to the coding style in this file, like `void Context::setAllowTF32CuBLAS(bool b)` and `void Context::setDeterministicMkldnn(bool b)`

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

EikanWang commented on aten/src/ATen/native/mkldnn/xpu/detail/Conv.cpp:
```diff
@@ -211,6 +213,12 @@ sycl::event convolution_backward_weights(
 #endif
 
+  auto& ctx = at::globalContext();
+  bool allow_tf32 = ctx.allowTF32Onednn();
```
Onednn -> OneDNN

yanbing-j replied:
In oneDNN github https://github.com/oneapi-src/oneDNN, it uses `oneDNN`. Is it better to all use `oneDNN`?

ZhiweiYan-96 replied:
hi, @EikanWang  @yanbing-j , I've changed `Onednn` to `oneDNN`. I think using the project name `oneDNN`  is easier for us to align naming style.

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

EikanWang commented on aten/src/ATen/Context.h:
```diff
@@ -320,6 +320,8 @@ class TORCH_API Context {
   void setFloat32MatmulPrecision(const std::string& s);
   bool allowTF32CuDNN() const;
   void setAllowTF32CuDNN(bool);
+  bool allowTF32oneDNN() const;
+  void setAllowTF32oneDNN(bool);
```
It should be `OneDNN` for code style consistency.

ZhiweiYan-96 replied:
Thank for suggestions! I have changed the name to `OneDNN` in PR,  I've already synced the change with Yanbing.

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

atalman reviewed (approved):
lgtm

atalman commented:
@pytorchmergebot merge -i

malfet commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.h:
```diff
@@ -80,7 +82,7 @@ dnnl::memory::dims compatible_dilation(Vec&& dilation) {
   }
   return ret;
-}
+};
```
Why this change is needed?
```suggestion
}
```

malfet commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.h:
```diff
@@ -94,6 +96,6 @@ dnnl::memory dnnl_memory_from_host_scalar(
   dnnl::memory mem = make_onednn_memory(md, engine, holder.data_ptr());
   return mem;
-}
+};
```
@Skylion007 Don't we have no-extra-semi linter now? 
```suggestion
}
```

malfet reviewed (changes requested):
IMO proposed API name is confusing:
 Either add `allowTF32Acceleration` that will enable it on both CUBLAS/CUDNN and XPU, or rename the call to `allowTF32XPU`, because oneDNN also covers CPUs, but this change has no effect on CPU, is it?

Nit: Please don't add semicolons after at the end of method/function implementation

EikanWang commented:
> IMO proposed API name is confusing: Either add `allowTF32Acceleration` that will enable it on both CUBLAS/CUDNN and XPU, or rename the call to `allowTF32XPU`, because oneDNN also covers CPUs, but this change has no effect on CPU, is it?

@malfet, Since oneDNN is a backend for both CPU and Intel GPU, `allowTF32OneDNN` will be applied to both CPU and Intel GPU. This PR is dedicated to Intel GPU. CPU will follow the same logic and submit another PR to enable the feature with the same user-facing API.

guangyey commented:
@ZhiweiYan-96 Please fix extra semicolons.

malfet commented:
> Since oneDNN is a backend for both CPUs and Intel GPU, `allowTF32OneDNN` will be applied to both CPUs and Intel GPU. This PR is dedicated to Intel GPU. CPU will follow the same logic and submit another PR to enable the feature with the same user-facing API.

This sounds reasonable. So, this functionality for CPU is already present in oneDNN and just needs to be exposed in PyTorch? Or will be added in the later version of oneDNN? If later, that I would still insist on an XPU name, because it might, or might not get added.

Alternatively, you can keep the name, but it should raise an error/print warning if one attemts to set it for PyTorch builds compiled without XPU support

EikanWang commented:
> So, this functionality for CPU is already present in oneDNN and just needs to be exposed in PyTorch?

@malfet , Yes, oneDNN has provided the feature for the CPU as well. 

> Alternatively, you can keep the name, but it should raise an error/print warning if one attempts to set it for PyTorch builds compiled without XPU support

@malfet Good point. @ZhiweiYan-96 , let's refine the PR and re-request review as long as the refinement is done.

ZhiweiYan-96 commented:
@EikanWang  @malfet  @guangyey  sure, thanks for your suggestions, let me refine the RP.

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 commented:
hi, @malfet  @EikanWang  Great thanks for suggestion, I have added the warning in commit https://github.com/pytorch/pytorch/pull/137570/commits/ab586cd566d3497fb11e37723121a559f83205f9 and fixed the format issue. Could you please have a look, thanks!

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

EikanWang commented:
The root cause of the CI failures is https://github.com/pytorch/pytorch/issues/142102, and we have submitted the PR https://github.com/pytorch/pytorch/pull/142113 to fix it.

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

EikanWang commented:
@pytorchbot rebase

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

etaf commented:
@pytorchbot  rebase

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

guangyey commented:
@malfet May I know if this PR is reasonable for you.

EikanWang commented:
@ZhiweiYan-96 , pls. stack the GEMM tf32 PR.

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 commented:
hi, @EikanWang, the code of tf32 is still in preparation. I will stack the code once I've finished the development, great thanks for reminding.

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

EikanWang commented:
@pytorchbot rebase

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 commented:
Hi, @malfet  @atalman  @ezyang @EikanWang , may I know this PR is reasonable for you? By the way, the only remaining CI error has a fix at [144238](https://github.com/pytorch/pytorch/pull/144238) already.

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 commented:
hi,  @atalman @malfet @ezyang  @EikanWang  just a quick ping to check if you’ve had a chance to look at this PR again. The recent commits add CPU during the setting in the cotext setting, users would get warning if they try to set allow_tf32 without Intel GPU.
 Let me know if there’s anything I should address. Great appreciation for your suggestions.

malfet commented:
>  just a quick ping to check if you’ve had a chance to look at this PR again. The recent commits add consideration on CPU during the tf32 context setting, resolving the discussion here [#137570 (review)](https://github.com/pytorch/pytorch/pull/137570#pullrequestreview-2474205267). 

Sorry, perhaps I'm missing something, but in that comment, I've asked to rename the API to something with XPU in its name, as right now this option does not work for CPU.
```cpp
id Context::setAllowTF32OneDNN(bool b){
#ifdef USE_XPU
  allow_tf32_onednn = b;
#else
  TORCH_WARN("TF32 acceleration on top of oneDNN is available for Intel GPUs. The current Torch version does not have Intel GPU Support.");
#endif
}
```

malfet reviewed (changes requested):
Please rename option to something that explicitly includes XPU in it's name (say `allowTF32XPU`) or extend support to CPU.

Also, would be nice to have a link to a doc of some sort explaining what this one does (because the test plan only validates that boolean flag can be set, which would work even if nothing really changes on the XPU side.)

ZhiweiYan-96 commented:
hi @malfet Sincere thanks for your feedback and response. 
We carefully considered your point during our earlier discussions. Please allow me to explain why we did not introduce XPU-only utilities. After internal deliberation with @EikanWang and CPU colleagues, we reached a consensus that this API is designed to control behaviors at the granularity of the DNNL library level, rather than at the device level. Consequently, the current PR does not differentiate between CPU and XPU. CPU colleagues would add CPU logic when they have requirements in same API.

My earlier explanation might not have been clear enough. I hope this detailed reasoning makes sense—do you think this is reasonable?

cc @EikanWang  @liangan1 Please correct me if wrong info conveyed.

EikanWang commented:
@malfet , As we discussed earlier in this PR - [Since oneDNN is a backend for both CPUs and Intel GPU, allowTF32OneDNN will be applied to both CPUs and Intel GPU. This PR is dedicated to Intel GPU. CPU will follow the same logic and submit another PR to enable the feature with the same user-facing API.](https://github.com/pytorch/pytorch/pull/137570#issuecomment-2513795272), this API intends to serve both CPU and Intel GPU. 

In terms of the doc, @ZhiweiYan-96 , pls. add oneDNN documentation to elaborate on the tf32.

ZhiweiYan-96 commented:
@malfet @EikanWang

TF32 is a data type that can improve performance during the training and inference of deep learning workloads. To facilitate this data type, oneDNN provides the fp_math model API to control the TF32 down-conversion from FP32 during convolution and GEMM kernel computations. You can find more details in the oneDNN documentation [here](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_fpmath_mode.html#the-floating-point-math-mode-attribute).

The current unit test only validates the boolean flag. I had extra testing on convolution (forward, backward data, and backward weight), and you can find the oneDNN verbose output in [the PR description](https://github.com/pytorch/pytorch/pull/137570#issue-2575134260). The field attr-fpmath:tf32 in the verbose output shows that the TF32 conversion has been successfully triggered.

I will also add more detailed descriptions to the PR description—thanks for your kind reminder!

ZhiweiYan-96 commented:
Hi @malfet @EikanWang , could you please take a look at this PR when you have some time? Let me know if you have any suggestions. Thanks!

EikanWang commented:
@ZhiweiYan-96 , it looks good to me and I have approved it. @malfet , may I know if your comments have been addressed?

guangyey commented:
@pytorchbot rebase

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

malfet reviewed (approved):
I still think that's a weird API, but if it's only exposed as MKLDNN context, than sure, why not

malfet commented:
@pytorchbot merge -f "Lint is green, XPU is red, but for rest of CI it's a no-op"

guangyey commented:
@pytorchbot rebase

guangyey commented:
“Help to resolve conflicts, try to land"
@pytorchbot merge

guangyey committed:
Update

[ghstack-poisoned]

guangyey commented:
“It seems unrelated. Try rebasing instead.."
@pytorchbot rebase

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

guangyey commented:
@ZhiweiYan-96 Please help ensure the failure is irrelevant to this PR.

EikanWang commented:
@pytorchbot merge -i

alexbaden commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.cpp:
```diff
@@ -1,5 +1,8 @@
 #include <ATen/native/ConvUtils.h>
 #include <ATen/native/mkldnn/xpu/detail/Utils.h>
+#include <dnnl.hpp>
+#include <dnnl_common.hpp>
```
@ZhiweiYan-96 this file `dnnl_common.hpp` is in the `oneapi/dnnl` subdirectory of `intel/oneapi/dnnl/latest/include/` which is not included by default in the oneapi include paths. On windows this is causing a problem. Can we update the path here to include `oneapi/dnnl` prefix? Do you want me to file an issue?
